File: common/src/main/java/com/codeloom/common/language/LanguageSpec.java
package com.codeloom.common.language;

import java.util.Arrays;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor
public enum LanguageSpec {
    JAVA("java", "eclipse-temurin:21-jdk", "Main.java", "javac Main.java", "java Main < input.txt"),
    CPP("cpp", "gcc:15", "main.cpp", "g++ -std=c++20 -O2 -o main main.cpp", "./main < input.txt"),
    PYTHON("python", "python:3.14-slim", "main.py", null, "python3 main.py < input.txt");

    private final String id;
    private final String image;
    private final String sourceFileName;
    private final String compileCommand;
    private final String runCommand;

    public static LanguageSpec fromLanguage(String language) {
        return Arrays.stream(values())
                .filter(v -> v.id.equalsIgnoreCase(language))
                .findFirst()
                .orElseThrow(() -> new InvalidLanguageException(language));
    }
}

File: common/src/main/kotlin/com/codeloom/common/language/LanguageSpec.kt
package com.codeloom.common.language

enum class LanguageSpec(
    val id: String,
    val image: String,
    val sourceFileName: String,
    val compileCommand: String? = null,
    val runCommand: String,
) {
    JAVA(
        id = "java",
        image = "eclipse-temurin:21-jdk",
        sourceFileName = "Main.java",
        compileCommand = "javac Main.java",
        runCommand = "java Main < input.txt",
    ),

    CPP(
        id = "cpp",
        image = "gcc:15",
        sourceFileName = "main.cpp",
        compileCommand = "g++ -std=c++20 -O2 -o main main.cpp",
        runCommand = "./main < input.txt",
    ),

    PYTHON(
        id = "python",
        image = "python:3.14-slim",
        sourceFileName = "main.py",
        compileCommand = null,
        runCommand = "python3 main.py < input.txt",
    ),
    ;

    companion object {
        private val byId = entries.associateBy(LanguageSpec::id)

        fun fromLanguage(language: String): LanguageSpec =
            byId[language.lowercase()]
                ?: throw InvalidLanguageException(language)
    }
}
